### PR TITLE
Groundwork to store validators also in a bags-list instance (2)

### PR DESCRIPTION
**Pull request by kianenigma**
This PR adds the groundwork to store validators in bags-list, sorted based on stake. This builds on top of #10821 and should be merged after it.

Similar to what we did with nominators, we can deploy this in multiple stages: 

1. current code can be deployed as is
2. Then, we can update staking to `Release::V10`, and migrate all validators to the bags-list, but use a hypothetical `UseValidatorsMapAndUpdateBagsList`.
3. Once confident, we switch to `type TargetList = ValidatorsBagsList`.

### CODE REVIEW

File: frame/multisig/src/lib.rs [hunk 1]
```diff
@@ -24,12 +24,11 @@
...
-//! origins (accounts) to coordinate and dispatch a call from a well-known origin, derivable
-//! deterministically from the set of account IDs and the threshold number of accounts from the
-//! set that must approve it. In the case that the threshold is just one then this is a stateless
-//! operation. This is useful for multisig wallets where cryptographic threshold signatures are
-//! not available or desired.
+//! operation, allowing multiple signed origins (accounts) to coordinate and dispatch a call from a
+//! well-known origin, derivable deterministically from the set of account IDs and the threshold
+//! number of accounts from the set that must approve it. In the case that the threshold is just one
+//! then this is a stateless operation. This is useful for multisig wallets where cryptographic
+//! threshold signatures are not available or desired.
```
Comment (emostov) [inline frame/multisig/src/lib.rs:31]: Looks like just a line re-wrap? IMO it would be preferable to keep this out since it seems totally unrelated

File: frame/staking/src/lib.rs [hunk 2]
```diff
@@ -767,11 +767,12 @@ enum Releases {
...
-	V6_0_0, // removal of all storage associated with offchain phragmen.
-	V7_0_0, // keep track of number of nominators / validators in map
-	V8_0_0, // populate `SortedListProvider`.
-	V9_0_0, // inject validators into `SortedListProvider` as well.
+	V5_0_0,  // blockable validators.
+	V6_0_0,  // removal of all storage associated with offchain phragmen.
+	V7_0_0,  // keep track of number of nominators / validators in map
+	V8_0_0,  // populate `VoterList`.
+	V9_0_0,  // inject validators into `NPoSVoteProvider` as well.
+	V10_0_0, // inject validator into `NPoSTargetList`.
```
Comment (emostov) [inline frame/staking/src/lib.rs:775]:
```suggestion
	V9_0_0,  // inject validators into `VoterList` as well.
	V10_0_0, // inject validator into `TargetList`.
```

File: frame/staking/src/migrations.rs [hunk 3]
```diff
@@ -22,45 +22,79 @@ use frame_support::traits::OnRuntimeUpgrade;
...
+				let _ = T::VoterList::on_insert(v.clone(), weight).map_err(|err| {
+					log!(warn, "failed to insert {:?} into VoterList: {:?}", v, err)
 				});
 			}
 
+			StorageVersion::<T>::put(Releases::V9_0_0);
 			T::BlockWeights::get().max_block
 		} else {
-			log!(warn, "InjectValidatorsIntoSortedListProvider being executed on the wrong storage version, expected Releases::V8_0_0");
+			log!(warn, "InjectValidatorsIntoVoterList being executed on the wrong storage version, expected Releases::V8_0_0");
```
Comment (emostov) [inline frame/staking/src/migrations.rs:40]:
```suggestion
			log!(warn, "InjectValidatorsIntoVoterList being attempted on the wrong storage version, expected Releases::V8_0_0");
```
maybe? Executed makes it sound like we did the migration anyways

File: frame/staking/src/migrations.rs [hunk 4]
```diff
@@ -22,45 +22,79 @@ use frame_support::traits::OnRuntimeUpgrade;
...
+				let weight = Pallet::<T>::vote_weight(&v);
+				let _ = T::TargetList::on_insert(v.clone(), weight).map_err(|err| {
+					log!(warn, "failed to insert {:?} into TargetList: {:?}", v, err)
+				});
+			}
+
+			StorageVersion::<T>::put(Releases::V10_0_0);
+			T::BlockWeights::get().max_block
+		} else {
+			log!(warn, "InjectValidatorsIntoTargetList being executed on the wrong storage version, expected Releases::V9_0_0");
```
Comment (emostov) [inline frame/staking/src/migrations.rs:84]:
```suggestion
			log!(warn, "InjectValidatorsIntoTargetList being attempted on the wrong storage version, expected Releases::V9_0_0");
```
Same rationale as above - although not really a big deal

File: frame/staking/src/migrations.rs [hunk 5]
```diff
@@ -22,45 +22,79 @@ use frame_support::traits::OnRuntimeUpgrade;
...
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<(), &'static str> {
+		ensure!(StorageVersion::<T>::get() == Releases::V9_0_0, "must upgrade linearly");
+		Ok(())
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn post_upgrade() -> Result<(), &'static str> {
+		ensure!(StorageVersion::<T>::get(), Releases::V10_0_0, "must upgrade linearly");
```
Comment (emostov) [inline frame/staking/src/migrations.rs:97]: We should `ensure!(Validators::count() == TargetList::count(), "each validator must be in the target list")`

Comment (emostov) [reply frame/staking/src/migrations.rs:97]: Also `T::TargetList::sanity_check().map_err(|_| "TargetList is not in a sane state.")?;

File: frame/staking/src/migrations.rs [hunk 6]
```diff
@@ -22,45 +22,79 @@ use frame_support::traits::OnRuntimeUpgrade;
...
+pub struct InjectValidatorsIntoTargetList<T>(sp_std::marker::PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for InjectValidatorsIntoTargetList<T> {
+	fn on_runtime_upgrade() -> Weight {
+		if StorageVersion::<T>::get() == Releases::V9_0_0 {
+			for (v, _) in Validators::<T>::iter() {
+				let weight = Pallet::<T>::vote_weight(&v);
+				let _ = T::TargetList::on_insert(v.clone(), weight).map_err(|err| {
+					log!(warn, "failed to insert {:?} into TargetList: {:?}", v, err)
+				});
+			}
```
Comment (emostov) [inline frame/staking/src/migrations.rs:79]: To avoid duplicate code we can instead use:

```rust
TargetList::unsafe_regenerate(				
    Validators::<T>::iter().map(|(id, _)| id),
	Pallet::<T>::weight_of_fn(),
)
```

There isn't any performance benefit though (originally there was, but that never made it into master because we refactored `insert_many` to just use the normal `insert`)

File: frame/staking/src/pallet/impls.rs [hunk 7]
```diff
@@ -744,22 +745,18 @@ impl<T: Config> Pallet<T> {
...
-	pub fn get_npos_targets() -> Vec<T::AccountId> {
-		let mut validator_count = 0u32;
-		let targets = Validators::<T>::iter()
-			.map(|(v, _)| {
-				validator_count.saturating_inc();
-				v
-			})
+	pub fn get_npos_targets(maybe_max_len: Option<usize>) -> Vec<T::AccountId> {
+		let targets = T::TargetList::iter()
+			.take(maybe_max_len.unwrap_or_else(Bounded::max_value))
```
Comment (emostov) [inline frame/staking/src/pallet/impls.rs:750]: `Bounded::max_value` is usize::MAX? This is a bit scary, especially since the point of using the bags list here is to allow it to be virtually unbounded. IMO we should either hardcode a fallback or have a configurable fallback value.

File: frame/staking/src/pallet/impls.rs [hunk 8]
```diff
@@ -863,6 +851,18 @@ impl<T: Config> Pallet<T> {
 			DispatchClass::Mandatory,
 		);
 	}
+
+	/// Perform all checks related to both [`Config::TargetList`] and [`Config::VoterList`].
+	#[cfg(debug_assertions)]
+	fn sanity_check_list_providers() {
```
Comment (emostov) [inline frame/staking/src/pallet/impls.rs:857]: 👍

File: frame/staking/src/pallet/impls.rs [hunk 9]
```diff
@@ -863,6 +851,18 @@ impl<T: Config> Pallet<T> {
 			DispatchClass::Mandatory,
 		);
 	}
+
+	/// Perform all checks related to both [`Config::TargetList`] and [`Config::VoterList`].
+	#[cfg(debug_assertions)]
```
Comment (emostov) [inline frame/staking/src/pallet/impls.rs:856]: Since this function is gated by `#[cfg(debug_assertions)]`, could we remove that same expression from the calls sites i.e. change

```rust
#[cfg(debug_assertions)]
Self::sanity_check_list_providers();
```

to simply

```rust
Self::sanity_check_list_providers();
```

File: frame/staking/src/pallet/impls.rs [hunk 10]
```diff
@@ -876,22 +876,19 @@ impl<T: Config> ElectionDataProvider for Pallet<T> {
...
 	}
 
 	fn targets(maybe_max_len: Option<usize>) -> data_provider::Result<Vec<T::AccountId>> {
-		let target_count = Validators::<T>::count();
-
-		// We can't handle this case yet -- return an error.
-		if maybe_max_len.map_or(false, |max_len| target_count > max_len as u32) {
-			return Err("Target snapshot too big")
+		let targets = Self::get_npos_targets(maybe_max_len);
+		if maybe_max_len.map_or(false, |m| targets.len() > m) {
```
Comment (emostov) [inline frame/staking/src/pallet/impls.rs:888]: If you take my suggestion: https://github.com/paritytech/substrate/pull/10944/files#r816353287, then we should also make sure to check that fallback the upper bound respected if `maybe_max_len` is `None`

File: frame/staking/src/pallet/impls.rs [hunk 11]
```diff
@@ -978,7 +975,7 @@ impl<T: Config> ElectionDataProvider for Pallet<T> {
 		<Validators<T>>::remove_all();
 		<Nominators<T>>::remove_all();
 
-		T::SortedListProvider::unsafe_clear();
+		T::VoterList::unsafe_clear();
```
Comment (emostov) [inline frame/staking/src/pallet/impls.rs:978]:
```suggestion
		T::VoterList::unsafe_clear();
		T::TargetList::unsafe_clear();
```

File: frame/staking/src/pallet/impls.rs [hunk 12]
```diff
@@ -1288,7 +1285,7 @@ impl<T: Config> VoteWeightProvider<T::AccountId> for Pallet<T> {
 	}
 }
 
-/// A simple voter list implementation that does not require any additional pallets. Note, this
+/// A simple sorted list implementation that does not require any additional pallets. Note, this
 /// does not provided nominators in sorted ordered. If you desire nominators in a sorted order take
```
Comment (emostov) [inline frame/staking/src/pallet/impls.rs:1289]:
```suggestion
/// does not provided voters in sorted ordered. If you desire voters in a sorted order take
```

File: frame/staking/src/pallet/impls.rs [hunk 13]
```diff
@@ -1331,9 +1328,50 @@ impl<T: Config> SortedListProvider<T::AccountId> for UseNominatorsAndValidatorsM
...
 	fn unsafe_clear() {
-		// NOTE: Caller must ensure this doesn't lead to too many storage accesses. This is a
-		// condition of SortedListProvider::unsafe_clear.
 		Nominators::<T>::remove_all();
 		Validators::<T>::remove_all();
 	}
 }
+
+/// A simple sorted list implementation that does not require any additional pallets. Note, this
+/// does not provided validators in sorted ordered. If you desire nominators in a sorted order take
```
Comment (emostov) [inline frame/staking/src/pallet/impls.rs:1337]:
```suggestion
/// does not provide validators in sorted ordered. If you desire nominators in a sorted order take
```

File: frame/staking/src/testing_utils.rs [hunk 14]
```diff
@@ -38,11 +38,11 @@ const SEED: u32 = 0;
 pub fn clear_validators_and_nominators<T: Config>() {
 	Validators::<T>::remove_all();
 
-	// whenever we touch nominators counter we should update `T::SortedListProvider` as well.
+	// whenever we touch nominators counter we should update `T::VoterList` as well.
 	Nominators::<T>::remove_all();
 
 	// NOTE: safe to call outside block production
-	T::SortedListProvider::unsafe_clear();
+	T::VoterList::unsafe_clear();
```
Comment (emostov) [inline frame/staking/src/testing_utils.rs:45]:
```suggestion
	T::VoterList::unsafe_clear();
	T::TargetList::unsafe_clear();
```

**Comment by emostov** [comment]
Check list for making sure target list (not voter list) is correctly called

|      extrinsic     |         Chilled        |        Nominator       | Validator              |
|:------------------:|:----------------------:|:----------------------:|------------------------|
| chill, chill_other | nothing, nothing       | nothing, nothing | on_remove,  dec counter       |
| nominate           | nothing, nothing | nothing, nothing       | on_remove, dec counter |
| validate           | nothing, nothing       | on_insert, inc counter | nothing, nothing       |
| bond_extra         | nothing, nothing       | nothing, nothing     | on_update, nothing       |
| rebond             | nothing, nothing       | nothing, nothing     | on_update, nothing       |
| unbond             | nothing, nothing       | nothing, nothing     | on_update, nothing       |

File: frame/staking/src/pallet/mod.rs [hunk 15]
```diff
@@ -816,10 +824,15 @@ pub mod pallet {
...
-				if T::SortedListProvider::contains(&stash) {
-					T::SortedListProvider::on_update(&stash, Self::weight_of(&ledger.stash));
-					debug_assert_eq!(T::SortedListProvider::sanity_check(), Ok(()));
+				if T::VoterList::contains(&stash) {
+					T::VoterList::on_update(&stash, Self::weight_of(&ledger.stash));
+					debug_assert_eq!(T::VoterList::sanity_check(), Ok(()));
+				}
+				if T::TargetList::contains(&stash) {
+					T::TargetList::on_update(&stash, Self::weight_of(&ledger.stash));
+					debug_assert_eq!(T::VoterList::sanity_check(), Ok(()));
```
Comment (emostov) [inline frame/staking/src/pallet/mod.rs:835]:
```suggestion
					debug_assert_eq!(T::TargetList::sanity_check(), Ok(()));
```
We also can remove the debug assert here and above and instead put `Self::sanity_check_list_providers();` after these two if statements

File: frame/staking/src/pallet/mod.rs [hunk 16]
```diff
@@ -816,10 +824,15 @@ pub mod pallet {
 
 				// NOTE: ledger must be updated prior to calling `Self::weight_of`.
 				Self::update_ledger(&controller, &ledger);
+
 				// update this staker in the sorted list, if they exist in it.
```
Comment (emostov) [inline frame/staking/src/pallet/mod.rs:828]:
```suggestion
				// update this staker in the relevant list
```

File: frame/staking/src/pallet/mod.rs [hunk 17]
```diff
@@ -885,8 +898,11 @@ pub mod pallet {
 				Self::update_ledger(&controller, &ledger);
 
 				// update this staker in the sorted list, if they exist in it.
```
Comment (emostov) [inline frame/staking/src/pallet/mod.rs:900]:
```suggestion
				// update this staker in the relevant list(s).
```

**Comment by kianenigma** [comment]
This will be re-opened again with a new version that keeps track of approval stakes, not self-stakes.
